src/main/java/uk/gov/hmcts/reform/sandl/snlevents/actions/session/DragAndDropSessionAction.java

```diff
@@ -0,0 +1,160 @@
+    private DragAndDropSessionRequest dragAndDropSessionRequest;
+    private SessionRepository sessionRepository;
+    private RoomRepository roomRepository;
+    private PersonRepository personRepository;
+    private EntityManager entityManager;
+    private Session session;
+    private List<HearingPart> hearingParts;
+    private String currentSessionAsString;
+
+    public DragAndDropSessionAction(DragAndDropSessionRequest dragAndDropSessionRequest,
+                                    SessionRepository sessionRepository,
+                                    RoomRepository roomRepository,
+                                    PersonRepository personRepository,
+                                    EntityManager entityManager,
+                                    ObjectMapper objectMapper) {
+        this.dragAndDropSessionRequest = dragAndDropSessionRequest;
+        this.sessionRepository = sessionRepository;
+        this.roomRepository = roomRepository;
+        this.personRepository = personRepository;
+        this.entityManager = entityManager;
+        this.objectMapper = objectMapper;
+    }
+
+    @Override
+    public void getAndValidateEntities() {
+        session  = sessionRepository.findOne(dragAndDropSessionRequest.getSessionId());
+        hearingParts = session.getHearingParts();
+        validateIfJudgeCanBeChanged(session, dragAndDropSessionRequest);
+    }
+
+    @Override
+    public void act() {
+        try {
+            currentSessionAsString = objectMapper.writeValueAsString(session);
+        } catch (JsonProcessingException e) {
+            throw new SnlEventsException("Given session couldn't be converted into string");
+        }
+
+        Session session = updateSession(dragAndDropSessionRequest);
+        entityManager.detach(session);
+        session.setVersion(dragAndDropSessionRequest.getVersion());
+
+        sessionRepository.save(session);
+    }
+
+    @Override
+    public List<FactMessage> generateFactMessages() {
+        SessionWithHearingPartsFacts sessionWithHpFacts =
+            factsMapper.mapDbSessionToRuleJsonMessage(session, hearingParts);
+        List<FactMessage> facts = sessionWithHpFacts.getHearingPartsFacts().stream()
+            .map(hpFact -> new FactMessage(RulesService.UPSERT_HEARING_PART, hpFact))
+            .collect(Collectors.toList());
+        facts.add(new FactMessage(RulesService.UPSERT_SESSION, sessionWithHpFacts.getSessionFact()));
+
+        return facts;
+    }
+
+    @Override
+    public List<UserTransactionData> generateUserTransactionData() {
+        List<UserTransactionData> userTransactionDataList = new ArrayList<>();
+        userTransactionDataList.add(new UserTransactionData("session",
+            session.getId(),
+            currentSessionAsString,
+            "update",
+            "update",
+            0)
+        );
+
+        for (HearingPart hp : hearingParts) {
+            String hearingPartAsString = null;
```

Comment:
Here you can also use UserTransactionDataPreparerService.mapHearingPartToStrings, you can find an example in WithdrawHearingAction